### PR TITLE
KnownProxies/Networks enhancements

### LINKED ISSUE #8417: KnownProxies/Networks enhancements
Following discussion on #2384.

Thus far ...

> Note that's an IPv4 address nested in an IPv6 address. Your KnownProxies or KnownNetworks would need to be represented as IPv6 addresses as well.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e548cd16-b06c-5802-a1a7-c627167857ae
* Version Independent ID: c48abb08-bea3-dcff-c067-2edf5dc9b3a6
* Content: [Configure ASP.NET Core to work with proxy servers and load balancers](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1)
* Content Source: [aspnetcore/host-and-deploy/proxy-load-balancer.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/proxy-load-balancer.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #8417 

* In addition to possibly providing more content for this, let me know if you have a good link for IPv4 in IPv6. I didn't see a MS doc link that looked good for it.
* Should we also note: https://docs.microsoft.com/dotnet/api/system.net.ipaddress.maptoipv6

### CODE REVIEW

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 1]
```diff
@@ -123,8 +123,8 @@ services.Configure<ForwardedHeadersOptions>(options =>
...
-| [KnownProxies](/dotnet/api/microsoft.aspnetcore.builder.forwardedheadersoptions.knownproxies) | Addresses of known proxies to accept forwarded headers from. Use `KnownProxies` to specify exact IP address matches.<br><br>The default is an [IList](/dotnet/api/system.collections.generic.ilist-1)\<[IPAddress](/dotnet/api/system.net.ipaddress)> containing a single entry for `IPAddress.IPv6Loopback`. |
+| [KnownNetworks](/dotnet/api/microsoft.aspnetcore.builder.forwardedheadersoptions.knownnetworks) | Address ranges of known networks to accept forwarded headers from. Provide IP ranges using Classless Interdomain Routing (CIDR) notation.<br><br>If a network address is an IPv4 address nested in an IPv6 address, supply the address to `KnownNetworks` in the nested IPv6 format. For more information, see the [Configuration for an IPv4 address represented as an IPv6 address](#configuration-for-an-ipv4-address-represented-as-an-ipv6-address) section.<br><br>The default is an [IList](/dotnet/api/system.collections.generic.ilist-1)\<[IPNetwork](/dotnet/api/microsoft.aspnetcore.httpoverrides.ipnetwork)> containing a single entry for `IPAddress.Loopback`. |
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: The nesting warrants an example like "::ffff:10.0.0.0"

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 2]
```diff
@@ -123,8 +123,8 @@ services.Configure<ForwardedHeadersOptions>(options =>
...
+| [KnownProxies](/dotnet/api/microsoft.aspnetcore.builder.forwardedheadersoptions.knownproxies) | Addresses of known proxies to accept forwarded headers from. Use `KnownProxies` to specify exact IP address matches.<br><br>If a proxy address is an IPv4 address nested in an IPv6 address, supply the address to `KnownProxies` in the nested IPv6 format. For more information, see the [Configuration for an IPv4 address represented as an IPv6 address](#configuration-for-an-ipv4-address-represented-as-an-ipv6-address) section.<br><br>The default is an [IList](/dotnet/api/system.collections.generic.ilist-1)\<[IPAddress](/dotnet/api/system.net.ipaddress)> containing a single entry for `IPAddress.IPv6Loopback`. |
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: Clarification: If the server is using dual mode sockets then IPv4 addresses will need to be supplied in an IPv6 format like "::ffff:10.0.0.1". See IPAddress.MapToIPv6. You can confirm if this is needed by looking at HttpContext.Connection.RemoteIpAddress.

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 3]
```diff
@@ -210,6 +210,36 @@ services.Configure<ForwardedHeadersOptions>(options =>
...
 
+### Configuration for an IPv4 address represented as an IPv6 address
+
+If a network or proxy address is an IPv4 address nested in an IPv6 address, supply the address to `KnownNetworks` or `KnownProxies` in the nested IPv6 format.
+
+In the following example, two network addresses that supply forwarded headers are added to the `KnownNetworks` in IPv6 format:
+
+* Long format IPv6 for `10.11.12.0/24`:
+  * `0000:0000:0000:0000:0000:ffff:0a0b:0c00`
+  * Prefix length: 24
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: The prefix length also needs to be adjusted for the IPv6 format. 24 -> 120, you've added 96 bits to the address.

**Comment by guardrex** [comment]
@Tratcher Is it ok that I added the clarification on dual-mode sockets to the section as opposed to the table? It seemed like a lot of text to have there, and it would be duplicated for the entries in the table. See the last commit.

**Comment by Tratcher** [comment]
I meant that as a replacement for the comment "If a network address is an IPv4 address nested in an IPv6 address (for example, `::ffff:10.0.0.0`), supply the address to `KnownNetworks` in the nested IPv6 format."

**Comment by guardrex** [comment]
This bit ...

> the last 32-bits provided in IPv4 format

Isn't correct. I'll provide another commit to generalize it.

**Comment by guardrex** [comment]
Would this (in the code example) ...

> The prefix length supplied to the IPNetwork constructor is the CIDR notation of the subnet mask.

... be better in either of the following formats ...

1. The IPv6 prefix length supplied to the IPNetwork constructor is the IPv4 CIDR notation of the subnet mask plus 96 [for example, the IPv4 CIDR subnet mask notation for 24 is 120 in IPv6 (24 + 96)].
1. The IPv6 prefix length supplied to the IPNetwork constructor is the IPv4 CIDR notation of the subnet mask in IPv6 format.

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 4]
```diff
@@ -210,6 +210,36 @@ services.Configure<ForwardedHeadersOptions>(options =>
 });
 ```
 
+### Configuration for an IPv4 address represented as an IPv6 address
+
+If a network or proxy address is an IPv4 address nested in an IPv6 address, supply the address to `KnownNetworks` or `KnownProxies` in the nested IPv6 format.
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: The same comment should be updated here.

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 5]
```diff
@@ -210,6 +210,36 @@ services.Configure<ForwardedHeadersOptions>(options =>
...
+
+In the following example, two network addresses that supply forwarded headers are added to the `KnownNetworks` in IPv6 format:
+
+* Long format IPv6 for `10.11.12.0/24`:
+  * `0000:0000:0000:0000:0000:ffff:0a0b:0c00`
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: No need to show the long format, most implementations show the short version.

File: aspnetcore/host-and-deploy/proxy-load-balancer.md (hunk 4 above)
Comment (Tratcher) [reply aspnetcore/host-and-deploy/proxy-load-balancer.md]: You could use a longer form here and abbreviate it in the tables if needed.

**Comment by guardrex** [comment]
@Tratcher Ok ... ready.

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 6]
```diff
@@ -210,6 +210,32 @@ services.Configure<ForwardedHeadersOptions>(options =>
 });
 ```
 
+### Configuration for an IPv4 address represented as an IPv6 address
+
+If the server is using dual-mode sockets, IPv4 addresses are supplied in an IPv6 format (for example, `10.0.0.1` in IPv4 represented in IPv6 as `::ffff:10.0.0.1` or `::ffff:a00:1`). See [IPAddress.MapToIPv6](xref:System.Net.IPAddress.MapToIPv6*). You can confirm if this format is required by looking at the [HttpContext.Connection.RemoteIpAddress](xref:Microsoft.AspNetCore.Http.ConnectionInfo.RemoteIpAddress*).
```
Comment (scottaddie) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: You can confirm --> Determine

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 7]
```diff
@@ -210,6 +210,32 @@ services.Configure<ForwardedHeadersOptions>(options =>
...
+
+In the following example, a network address that supplies forwarded headers is added to the `KnownNetworks` in IPv6 format:
```
Comment (scottaddie) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: the `KnownNetworks` --> the `KnownNetworks` list

**Comment by Tratcher** [comment]
Amusingly someone has taken it upon themselves to solve this another way: https://github.com/aspnet/BasicMiddleware/issues/358

File: aspnetcore/host-and-deploy/proxy-load-balancer.md [hunk 8]
```diff
@@ -210,6 +210,32 @@ services.Configure<ForwardedHeadersOptions>(options =>
...
 
+### Configuration for an IPv4 address represented as an IPv6 address
+
+If the server is using dual-mode sockets, IPv4 addresses are supplied in an IPv6 format (for example, `10.0.0.1` in IPv4 represented in IPv6 as `::ffff:10.0.0.1` or `::ffff:a00:1`). See [IPAddress.MapToIPv6](xref:System.Net.IPAddress.MapToIPv6*). Determine if this format is required by looking at the [HttpContext.Connection.RemoteIpAddress](xref:Microsoft.AspNetCore.Http.ConnectionInfo.RemoteIpAddress*).
+
+In the following example, a network address that supplies forwarded headers is added to the `KnownNetworks` list in IPv6 format:
+
+Short format IPv6 for `10.11.12.1/8`:
+
+* `::ffff:0a0b:0c01`
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/proxy-load-balancer.md]: `::ffff:10.11.12.1`

Comment (guardrex) [reply aspnetcore/host-and-deploy/proxy-load-balancer.md]: I think so ... There's already the little note for +96, so it's an easy matter to note down there that the hex address works, too. I'll ping u on the PR.

**Pull request merged**